what4/src/What4/Expr/Allocator.hs

```diff
@@ -0,0 +1,197 @@
+  mkOpt
+    cacheTerms
+    (cacheOptStyle gen storageRef szSetting)
+    (Just "Use hash-consing during term construction")
+    (Just (ConcreteBool False))
+
+
+------------------------------------------------------------------------
+-- | ExprAllocator provides an interface for creating expressions from
+-- an applications.
+-- Parameter @t@ is a phantom type brand used to track nonces.
+data ExprAllocator t
+   = ExprAllocator { appExpr  :: forall tp
+                            .  ProgramLoc
+                            -> App (Expr t) tp
+                            -> AbstractValue tp
+                            -> IO (Expr t tp)
+                  , nonceExpr :: forall tp
+                             .  ProgramLoc
+                             -> NonceApp t (Expr t) tp
+                             -> AbstractValue tp
+                             -> IO (Expr t tp)
+                  }
+
+
+------------------------------------------------------------------------
+-- Uncached storage
+
+-- | Create a new storage that does not do hash consing.
+newStorage :: NonceGenerator IO t -> IO (ExprAllocator t)
+newStorage g = do
+  return $! ExprAllocator { appExpr = uncachedExprFn g
+                         , nonceExpr = uncachedNonceExpr g
+                         }
+
+uncachedExprFn :: NonceGenerator IO t
+              -> ProgramLoc
+              -> App (Expr t) tp
+              -> AbstractValue tp
+              -> IO (Expr t tp)
+uncachedExprFn g pc a v = do
+  n <- freshNonce g
+  return $! mkExpr n pc a v
+
+uncachedNonceExpr :: NonceGenerator IO t
+                 -> ProgramLoc
+                 -> NonceApp t (Expr t) tp
+                 -> AbstractValue tp
+                 -> IO (Expr t tp)
+uncachedNonceExpr g pc p v = do
+  n <- freshNonce g
+  return $! NonceAppExpr $ NonceAppExprCtor { nonceExprId = n
+                                          , nonceExprLoc = pc
+                                          , nonceExprApp = p
+                                          , nonceExprAbsValue = v
+                                          }
+
+------------------------------------------------------------------------
+-- Cached storage
+
+cachedNonceExpr :: NonceGenerator IO t
+               -> PH.HashTable PH.RealWorld (NonceApp t (Expr t)) (Expr t)
+               -> ProgramLoc
+               -> NonceApp t (Expr t) tp
+               -> AbstractValue tp
+               -> IO (Expr t tp)
+cachedNonceExpr g h pc p v = do
+  me <- stToIO $ PH.lookup h p
+  case me of
+    Just e -> return e
+    Nothing -> do
+      n <- freshNonce g
+      let e = NonceAppExpr $ NonceAppExprCtor { nonceExprId = n
+                                            , nonceExprLoc = pc
+                                            , nonceExprApp = p
+                                            , nonceExprAbsValue = v
+                                            }
+      seq e $ stToIO $ PH.insert h p e
+      return $! e
```

Comment:
Is the use of `$!` here actually necessary? The line above it calls `seq e`, after all.

Reply:
good question; I don't think so